### PR TITLE
Add SameSite support to Cookie

### PR DESCRIPTION
**Pull request by amygbui**
# Description
Add SameSite support to the Cookie class with a default value of "lax" when not explicitly set, which is the default browser behavior according to the MDN spec.

MDN Docs: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie

# Motivation
$Pay needs to set different SameSite policies on staging vs production. Right now we're appending `samesite=lax` to the stringified cookie, however built-in support would be much nicer.

### CODE REVIEW

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (JakeWharton) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:353]: Since this type isn't exposed anywhere, there's no need to add additional allocations to ensure it's immutable to external manipulation.
```suggestion
      val validSameSites = listOf("lax", "strict", "none")
```

**Review by JakeWharton — COMMENTED** [review]
Mostly LGTM, but will leave actual approval to a more active maintainer

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:264]: Omit if `"None"`

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:286]: This is documented as `Lax` with a capital `L` ??

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (yschimke) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:112]: Is String the right type for this?  It may be is we want to handle even invalid values.  But figured worth asking.

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:286]: +1 - seems like we should prefer the typical case.

Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:286]: I believe the default should be `"None"`

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:110]:
```suggestion
   * Returns a string describing whether this cookie is sent for cross-site calls.
   *
   * Two URLs are on the same site if they share a [top private domain][HttpUrl.topPrivateDomain].
   * Otherwise, they are cross-site URLs.
   *
   * When a URL is requested, it may be in the context of another URL.
   *
   *  * **Embedded resources like images and iframes** in browsers use the context as the page in
   *    the address bar and the subject is the URL of an embedded resource.
   *
   *  * **Potentially-destructive navigations such as HTTP POST calls** use the context as the page
   *    originating the navigation, and the subject is the page being navigated to.
   *
   * The values of this attribute determine whether this cookie is sent for cross-site calls:
   *
   *  - "Strict": the cookie is omitted when the subject URL is an embedded resource or a
   *    potentially-destructive navigation.
   *
   *  - "Lax": the cookie is omitted when the subject URL is an embedded resource. It is sent for
   *    potentially-destructive navigation. This is the default value.
   *
   *  - None: the cookie is always sent. The "Secure" attribute must also be set when setting this
   *    value.
```

I wanna read the spec to confirm my understanding is correct. Also I wanna get the right terms; I don’t think 'Context' and 'Subject' are right, but there probably are words that are.

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:351]: `sameSite`

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:353]: We can’t be this strict here. New values may come before OkHttp is updated.

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:439]: default to `"lax"` here so we don’t need to fix it later?

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:539]: I don’t think we should do this. The parsed `Cookie`’s job is to represent what was encoded only. If the user wants to break the rules, that’s not our business at this layer.

**Review by swankjesse — COMMENTED** [review]
I think the test case needs updates. I’d like to read the spec a bit more before landing this.

We might also want to offer APIs to actually implement the behavior for this. Not exactly sure where though, because we don’t currently have a 'context' URL for outbound calls.

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (yschimke) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:112]: Closing based on other comments.

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (amygbui) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt]: Updated, though I'm concerned the default may change in the future and we'll be providing an incorrect default.

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (amygbui) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:364]: What is the recommendation? Don't validate what's passed in, or add a warning log that says the value passed in is not a recognized value? For parsing, I think it makes sense to parse whatever fields people set even if it breaks the rules. Do we apply the same philosophy for cookie construction?

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (amygbui) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:275]: If someone intentionally set the value to `"Lax"`, it seems unexpected to omit it from the cookie string.

Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:275]: I think we should make `sameSite` nullable so we can distinguish between explicitly setting `Lax` vs. not

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:297]: ... and I think the default should be null, not `Lax`

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:370]:
```suggestion
      if (sameSite.equalsIgnoreCase("None")) {
```

This avoids an allocation

File: okhttp/src/jvmTest/java/okhttp3/CacheTest.java
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/CacheTest.java:2018]: revert, let’s make `null` the default

File: okhttp/src/jvmTest/java/okhttp3/CookieTest.kt
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/CookieTest.kt:1]: Please add tests for parsing and building a cookie with various SameSite attributes, including absent, normal, and unexpected

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt]: Yeah my bad. Because we’re going to differentiate between absent and Lax I think we should make absent the default.

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (amygbui) [reply okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt]: Updated using Kotlin syntax: `String.equals(value, ignoreCase = true)`

File: okhttp/src/jvmTest/java/okhttp3/CookieTest.kt
Comment (amygbui) [inline okhttp/src/jvmTest/java/okhttp3/CookieTest.kt:570]: I want to call out that passing an empty string in is a valid operation. Should we require a non-empty string? This issue also applies to the `name` and `value` fields.

File: okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Cookie.kt:372]:
```suggestion
```

I’m anxious about this 'cause I’d like this to work for all cookies:

```kotlin
val cookie: Cookie = ...
val copy = cookie.newBuilder().build()
```

In this case we’re strict on `build()` in a way that we can’t / shouldn’t be on `parse()`.

File: okhttp/src/jvmTest/java/okhttp3/CookieTest.kt
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/CookieTest.kt:553]: this works for me.

File: okhttp/src/jvmTest/java/okhttp3/CookieTest.kt
Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/CookieTest.kt:570]: No, I don’t think empty strings are special.

File: okhttp/src/jvmTest/java/okhttp3/CookieTest.kt
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/CookieTest.kt:601]:
```suggestion
    /** Note that we permit building a cookie that doesn’t follow the rules. */
    @Test fun builderSameSiteNoneDoesNotRequire() {
```

File: okhttp/src/jvmTest/java/okhttp3/CookieTest.kt
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/CookieTest.kt:610]:
```suggestion
```

File: okhttp/src/jvmTest/java/okhttp3/CookieTest.kt
Comment (swankjesse) [inline okhttp/src/jvmTest/java/okhttp3/CookieTest.kt:612]:
```suggestion
    val cookie = cookieBuilder.build()
```

**Review by swankjesse — APPROVED** [review]
Nice tests! 🚀

**Pull request merged**
